design-documents/mde-project-structure.md

```diff
@@ -0,0 +1,41 @@
+# Overview
+
+There is a need to have unified approach to structuring the Git-based project while developing MDEs (Magento Developed Extensions), like Page Builder and MSI.
+
+Currently, MSI is developed in a fork of main Magento repository, while Page Builder is developed in a separate repository (like B2B).
+
+## MDE project structure for Git-based installations
+
+MDE project must be structured as Magento B2B:
+ 1. Modules must be placed under `app/code`
+ 1. Each module must have valid `composer.json` with no version specified. Package version must be generated and added to the `composer.json` during release publication
+ 1. No core modules must be committed to the MDE repository
+ 1. Unit and MFTF tests must be distributed between modules (since they are modular). All other tests must be put inside `dev/tests` like it is done in Magento Open Source repository (because these testing frameworks do not support modularity)
+
+## Git-based project linking for local development
+
+Preferred way of linking the project is using [Composer path repository](https://getcomposer.org/doc/05-repositories.md#path).
+
+1. Check out magento 2 open source edition from Git into `<project_root>/magento2ce`
+1. Check out MDE into `<project_root>/magento2ce/<mde-name>`
+1. Add path composer repository declaration to `<project_root>/magento2ce/composer.json`:
```

Comment:
It will create bad developer experience. If contributor clones both core and MDE and would like to contribute in both she will have modified uncommited changes in composer.json file. Composer.lock file also should be managed separately. I believe we will have a tons of cases when composer.kson and composer.lock files will be modified in PRs unintentional.

Reply:
Any suggestions on alternative approach to prevent changes in composer.json?

Reply:
As described in another comment. We might keep config in separate file ignored by Git and apply it by composer plugin or independent CLI tool (extract linker from Magento commerce and allow to read config files instead of passing everything from command arguments). To resolve an issue with dependencies of MDE we may utilize https://github.com/wikimedia/composer-merge-plugin

Reply:
This seems to be overcomplicated and not transparent. Alternative approach is to commit dependencies on all MDEs in the default composer.json in magento2ce/magento2ee. 
Then the developer will have to checkout all MDEs along with the main repository to start development.